### PR TITLE
Cache restore payload paths in redis

### PR DESCRIPTION
**Pull request by benrudolph**
@snopoke this refactors restores to cache restore payload paths in redis. best read 🐟 

_Previous world_
On initial sync we would cache the payload path in redis if it took a long time. When the user was in steady state sync, we would store the cache path on the sync log token

_New world_
All syncs store the path to the cached restore in redis

cc: @calellowitz

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/phone/tests/test_ota_restore.py [hunk 1]
```diff
@@ -131,6 +131,11 @@ def testOverwriteCache(self):
         self.assertIsInstance(restore_config_cached.get_payload(), CachedResponse)
         self.assertNotIsInstance(restore_config_overwrite.get_payload(), CachedResponse)
 
+        # even cached responses change the sync log id so they are not the same
```
Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_ota_restore.py]: here is the crux of the `finalize` problem in the `CachedPayload`. this will continue to fail because we don't replace the sync log id in the restore. i really want to avoid having to do that. i think we discussed that we can just add the device_id to the restore_cache_key and that will alleviate the issue (have not done that yet). it's a bit confusing to me still cause i'm not exactly sure why it's bad when a user is logged into multiple devices and gets the same restore payload

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_logs.py [hunk 2]
```diff
@@ -78,29 +78,6 @@ def test_archive(self):
         self.assertEqual(0, len(log.get_footprint_of_cases_on_phone()))
 
 
-class CachingResponseTest(TestCase):
-
-    def testCachingResponse(self):
```
Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_logs.py]: i no longer think this test is necessary

**Comment by benrudolph** [comment]
I think this is g2g barring this one question here: https://github.com/dimagi/commcare-hq/pull/15884#discussion_r113209497. That determines how I fix the last failing test

File: corehq/ex-submodules/casexml/apps/phone/tests/test_ota_restore.py (hunk 1 above)
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_ota_restore.py]: It's bad because then their submissions from different forms will be attributed to the same synclog and the state will diverge from what is on the phone causing the phone to 412.

Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_ota_restore.py]: It is an edge case which probably doesn't happen that often (though I don't have data to support that assumption). If we include the device ID in the key then I think we can go ahead.

Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_ota_restore.py]: > It's bad because then their submissions from different forms will be attributed to the same synclog and the state will diverge from what is on the phone causing the phone to 412.

Ah! that was the lightbulb I needed 💡. thanks, will update with device id

Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_ota_restore.py]: https://github.com/dimagi/commcare-hq/pull/15884/commits/bfbb1e37a02767ac64e513bae9ef3c993186b29c

File: corehq/ex-submodules/casexml/apps/phone/restore.py [hunk 3]
```diff
@@ -436,14 +407,22 @@ class RestoreParams(object):
...
+    :param device_id:           The Device id of the device restoring
     """
 
-    def __init__(self, sync_log_id='', version=V1, state_hash='', include_item_count=False, app=None):
+    def __init__(self,
+            sync_log_id='',
+            version=V1,
+            state_hash='',
+            include_item_count=False,
+            device_id=None,
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/restore.py]: not seeing where this gets passed in from the view

Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/phone/restore.py]: forgot to commit the `views.py` file :man_facepalming:

Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/restore.py]: code review ftw :grin:

**Review by snopoke — APPROVED** [review]

File: corehq/ex-submodules/casexml/apps/phone/restore.py (hunk 3 above)
Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/phone/restore.py]: eagle eyes 🐦
